core/src/main/java/org/elasticsearch/snapshots/Snapshot.java

```diff
@@ -277,9 +293,14 @@ public static Snapshot fromXContent(XContentParser parser) throws IOException {
         int totalShard = 0;
         int successfulShards = 0;
         List<SnapshotShardFailure> shardFailures = NO_FAILURES;
-
-        XContentParser.Token token = parser.currentToken();
-        if (token == XContentParser.Token.START_OBJECT) {
+        if (parser.currentToken() == null) { // fresh parser? move to the first token
+            parser.nextToken();
+        }
+        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {  // on a start object move to next token
+            parser.nextToken();
+        }
+        XContentParser.Token token;
+        if ((token = parser.nextToken()) == XContentParser.Token.START_OBJECT) {
```

Comment:
No, we are looking for `{ "snapshot": { .... } }`. First we make sure that we came from the fresh parser, during this check the current token is null. So we set it to the first `{`. The second if verifies that we have got this `{` and moves parser once step further. So, now the current token is field name `snapshot`. The third if moves parser to the second `{` with `parser.currentName()` still being `snapshot`. I hope it makes sense. This structure enables us to use this parser in case we just opened a new file and in case we found it inside  a bigger XContent like
```
{
    "some_parent": {
        "something": {
        },
        "snapshot": {
           ......
        }
    }
}
```